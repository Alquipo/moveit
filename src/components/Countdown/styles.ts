import styled, { css } from 'styled-components'

export const WrapperCountdown = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    font-family: ${theme.font.family.rajdhani};
    font-weight: ${theme.font.bold};
    color: ${theme.colors.title};

    div {
      flex: 1;

      display: flex;
      align-items: center;
      justify-content: space-evenly;

      background-color: ${theme.colors.backgroundBox};
      transition: background-color 0.3s ease-in-out, color 0.3s ease-in-out;

      box-shadow: 0 0 60px rgba(0, 0, 0, 0.05);
      border-radius: 5px;
      font-size: 8.5rem;
      text-align: center;

      span {
        flex: 1;
      }

      span:first-child {
        border-right: 1px solid ${theme.colors.background};
        transition: border-right 0.3s ease-in-out;
      }
      span:last-child {
        border-left: 1px solid ${theme.colors.background};
        transition: border-left 0.3s ease-in-out;
      }
    }

    & > span {
      font-size: 6.25rem;
      margin: 0 0.5rem;
    }
  `}
`
export const CountdownButton = styled.button`
  ${({ theme }) => css`
    width: 100%;
    height: 5rem;

    margin-top: 2rem;

    display: flex;
    align-items: center;
    justify-content: center;

    border: 0;
    border-radius: 5px;

    background-color: ${theme.colors.blue};

    color: ${theme.colors.backgroundBox};

    font-size: 1.25rem;
    font-weight: ${theme.font.bold};

    transition: background-color 0.3s ease-in-out, color 0.3s ease-in-out;

    outline: 0;

    &:hover {
      background-color: ${theme.colors.blueDark};
    }
  `}
`
