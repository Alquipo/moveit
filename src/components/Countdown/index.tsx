import { useContext } from 'react'
import { CountdownContext } from 'contexts/CountdownContext'

import { Close } from '@styled-icons/evaicons-solid/Close'
import { Play } from '@styled-icons/foundation/Play'
import { CheckCircle } from '@styled-icons/boxicons-solid/CheckCircle'

import * as S from './styles'

const Countdown = () => {
  const {
    minutes,
    seconds,
    hasFinished,
    isActive,
    resetCountdown,
    startCountdown
  } = useContext(CountdownContext)

  const [minuteLeft, minuteRight] = String(minutes).padStart(2, '0').split('')

  const [secondsLeft, secondsRight] = String(seconds).padStart(2, '0').split('')

  return (
    <>
      <S.WrapperCountdown>
        <div>
          <span>{minuteLeft}</span>
          <span>{minuteRight}</span>
        </div>
        <span>:</span>
        <div>
          <span>{secondsLeft}</span>
          <span>{secondsRight}</span>
        </div>
      </S.WrapperCountdown>

      {hasFinished ? (
        <S.CountdownButton disabled>
          Ciclo Encerrado <CheckCircle />
        </S.CountdownButton>
      ) : (
        <>
          {isActive ? (
            <S.CountdownButton isActive type="button" onClick={resetCountdown}>
              Abandonar um ciclo <Close />
            </S.CountdownButton>
          ) : (
            <S.CountdownButton type="button" onClick={startCountdown}>
              Iniciar um ciclo <Play />
            </S.CountdownButton>
          )}
        </>
      )}
    </>
  )
}

export default Countdown
