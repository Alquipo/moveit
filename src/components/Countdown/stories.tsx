import { Story, Meta } from '@storybook/react/types-6-0'
import Countdown from '.'

export default {
  title: 'Countdown',
  component: Countdown
} as Meta

export const Default: Story = (args) => (
  <div style={{ maxWidth: '400px', margin: '0 auto' }}>
    <Countdown {...args} />
  </div>
)
