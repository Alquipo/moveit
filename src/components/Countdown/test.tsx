import { fireEvent, screen } from '@testing-library/react'

import { renderWithTheme } from 'utils/tests/helpers'
import Countdown from '.'

describe('<Countdown />', () => {
  it('should render the heading', () => {
    const { container } = renderWithTheme(<Countdown />)

    expect(screen.getByText(/2/i)).toBeInTheDocument()
    expect(screen.getByText(/5/i)).toBeInTheDocument()
    expect(screen.getAllByText(/0/i)).toHaveLength(2)

    expect(
      screen.getByRole('button', { name: /iniciar um ciclo/i })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  jest.useFakeTimers()
  it('should call counter press button', () => {
    renderWithTheme(<Countdown />)

    fireEvent.click(screen.getByRole('button'))

    expect(setTimeout).toHaveBeenCalledTimes(2)
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 1000)
  })
})
