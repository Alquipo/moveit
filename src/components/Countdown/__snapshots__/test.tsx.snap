// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Countdown /> should render the heading 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: "Rajdhani",sans-serif;
  font-weight: 600;
  color: #FFFFFF;
}

.c0 div {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  background-color: #202024;
  -webkit-transition: background-color 0.3s ease-in-out,color 0.3s ease-in-out;
  transition: background-color 0.3s ease-in-out,color 0.3s ease-in-out;
  box-shadow: 0 0 60px rgba(0,0,0,0.05);
  border-radius: 5px;
  font-size: 8.5rem;
  text-align: center;
}

.c0 div span {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 div span:first-child {
  border-right: 1px solid #121214;
  -webkit-transition: border-right 0.3s ease-in-out;
  transition: border-right 0.3s ease-in-out;
}

.c0 div span:last-child {
  border-left: 1px solid #121214;
  -webkit-transition: border-left 0.3s ease-in-out;
  transition: border-left 0.3s ease-in-out;
}

.c0 > span {
  font-size: 6.25rem;
  margin: 0 0.5rem;
}

<div
  class="c0"
>
  <div>
    <span>
      2
    </span>
    <span>
      5
    </span>
  </div>
  <span>
    :
  </span>
  <div>
    <span>
      0
    </span>
    <span>
      0
    </span>
  </div>
</div>
`;
